---
title: "Steps to get read count"
author: "Runan Yao"
date: "September 25, 2018"
output: html_document
---


This file discribe each step of getting the read count data we used in last several week.
Based on WorkStation Server.

Gene2 one can found in https://github.com/HalforcNull/STATBioinformatics/blob/master/HW%205%20Gene2%20version.Rmd

# 0. Operating System

Ubuntu 

# 1. Get data - Download data using fastq-dump

```{bash eval=FALSE}
nohup fastq-dump --gzip --split-files SRR493366 &
nohup fastq-dump --gzip --split-files SRR493367 &
nohup fastq-dump --gzip --split-files SRR493368 &
nohup fastq-dump --gzip --split-files SRR493369 &
nohup fastq-dump --gzip --split-files SRR493370 &
nohup fastq-dump --gzip --split-files SRR493371 &
```

This is download process need about 2.5 hours.

Instead, we can link the files if they are already downloaded in some folder.

# 2. Get data - Link downloaded data

Workstation already having all files saved in /STAT736 folder.
What we need to do is create our own folder, and link files from STAT736 to our data folder.

## 2.1 Create your own data folder

```{bash eval=FALSE}
mkdir mydata
cd mydata
```


## 2.3 link from data folder

```{bash eval=FALSE}
ln -s /STAT736/SRR493366_1.fastq.gz SRR493366_1.fastq.gz
ln -s /STAT736/SRR493366_2.fastq.gz SRR493366_2.fastq.gz
ln -s /STAT736/SRR493367_1.fastq.gz SRR493367_1.fastq.gz
ln -s /STAT736/SRR493367_2.fastq.gz SRR493367_2.fastq.gz
ln -s /STAT736/SRR493368_1.fastq.gz SRR493368_1.fastq.gz
ln -s /STAT736/SRR493368_2.fastq.gz SRR493368_2.fastq.gz
ln -s /STAT736/SRR493369_1.fastq.gz SRR493369_1.fastq.gz
ln -s /STAT736/SRR493369_2.fastq.gz SRR493369_2.fastq.gz
ln -s /STAT736/SRR493370_1.fastq.gz SRR493370_1.fastq.gz
ln -s /STAT736/SRR493370_2.fastq.gz SRR493370_2.fastq.gz
ln -s /STAT736/SRR493371_1.fastq.gz SRR493371_1.fastq.gz
ln -s /STAT736/SRR493371_2.fastq.gz SRR493371_2.fastq.gz
```

# 3. fastqc

Fastqc will generate a report based on fastq read quality. 
https://en.wikipedia.org/wiki/FASTQ_format

```{bash eval=FALSE}
fastqc -q SRR493366_1.fastq.gz &
fastqc -q SRR493366_2.fastq.gz &
fastqc -q SRR493367_1.fastq.gz &
fastqc -q SRR493367_2.fastq.gz &
fastqc -q SRR493368_1.fastq.gz &
fastqc -q SRR493368_2.fastq.gz &
fastqc -q SRR493369_1.fastq.gz &
fastqc -q SRR493369_2.fastq.gz &
fastqc -q SRR493370_1.fastq.gz &
fastqc -q SRR493370_2.fastq.gz &
fastqc -q SRR493371_1.fastq.gz &
fastqc -q SRR493371_2.fastq.gz &
```

# 4. Kallisto

## 4.1 Download cdna data and ncrna data

Since we are discussing human cancer, we are download cdna and ncrna of homo sapiens.

After download, we combine them into one data file.
```{bash eval=FALSE}
wget ftp://ftp.ensembl.org/pub/current_fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz
wget ftp://ftp.ensembl.org/pub/current_fasta/homo_sapiens/ncrna/Homo_sapiens.GRCh38.ncrna.fa.gz
cat Homo_sapiens.GRCh38.cdna.all.fa.gz Homo_sapiens.GRCh38.ncrna.fa.gz > Homo_sapiens.GRCh38.rna.fa.gz
```

## 4.2 Build index

```{bash eval=FALSE}
kallisto index -i hsGRCh38_kallisto Homo_sapiens.GRCh38.rna.fa.gz
```

## 4.3 Run kallisto

-i hsGRCh38_kallisto : the index we just build
-t 4 : use 4 core

If this is single end read, we need setup the read length using: -l 

```{bash eval=FALSE}
kallisto quant -i hsGRCh38_kallisto -t 4 -o SRR493366 SRR493366_1.fastq.gz SRR493366_2.fastq.gz &
kallisto quant -i hsGRCh38_kallisto -t 4 -o SRR493367 SRR493367_1.fastq.gz SRR493367_2.fastq.gz &
kallisto quant -i hsGRCh38_kallisto -t 4 -o SRR493368 SRR493368_1.fastq.gz SRR493368_2.fastq.gz &
kallisto quant -i hsGRCh38_kallisto -t 4 -o SRR493369 SRR493369_1.fastq.gz SRR493369_2.fastq.gz &
kallisto quant -i hsGRCh38_kallisto -t 4 -o SRR493370 SRR493370_1.fastq.gz SRR493370_2.fastq.gz &
kallisto quant -i hsGRCh38_kallisto -t 4 -o SRR493371 SRR493371_1.fastq.gz SRR493371_2.fastq.gz &
```

# 5. Convert read into one table using R

```{r eval=FALSE}
library(EnsDb.Hsapiens.v86)

esdb <- EnsDb.Hsapiens.v86

newtxs <- transcripts(esdb, return.type = 'data.frame')
k <- keys(esdb, keytype = "TXNAME")
tx2gene <- dplyr::select(newtxs, one_of(c('tx_name', 'gene_id')))
colnames(tx2gene) <- c('TXNAME', 'GENEID')

files <- c(
  'SRR493366/abundance.tsv',
  'SRR493367/abundance.tsv',
  'SRR493368/abundance.tsv',
  'SRR493369/abundance.tsv',
  'SRR493370/abundance.tsv',
  'SRR493371/abundance.tsv')
names(files) <- c('Scramble1','Scramble2','Scramble3','HOXA1KD1','HOXA1KD2','HOXA1KD3')

txi.kallisto.tsv <- tximport(files, type = "kallisto", tx2gene = tx2gene, ignoreAfterBar = TRUE, ignoreTxVersion=TRUE )
write.csv(txi.kallisto.tsv$counts, file="ReadCount(Kallisto).csv", row.names = TRUE)


```
